packages/react-i18n/src/index.tsx

```diff
@@ -0,0 +1,102 @@
+/**
+ * External dependencies
+ */
+import * as React from 'react';
+import { __, _n, _nx, _x, setLocaleData } from '@wordpress/i18n';
+
+export interface I18nProps {
+	__: typeof __;
+	_n: typeof _n;
+	_nx: typeof _nx;
+	_x: typeof _x;
+	i18nLocale?: string;
+}
+
+const I18nContext = React.createContext< I18nProps >( { __, _n, _nx, _x } );
+
+interface Props {
+	/**
+	 * The current locale
+	 */
+	locale: string;
+
+	/**
+	 * A callback that resolves with the translations data
+	 */
+	onLocaleChange?( newLocale: string | undefined ): Promise< object >;
```

Comment:
The `onLocaleChange` name can be confusing. Typically, the `locale` and `onLocaleChange` pair would be used for a controlled component where an user interaction (or other external event) can change the value:
```js
<input type="locale" locale={ this.state.locale } onLocaleChange={ locale => this.setState( { locale } ) } />
```

But this function does something very different. It implements an incomplete interface. Maybe `loadLocaleData` would be a better name?